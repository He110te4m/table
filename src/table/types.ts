/* eslint-disable @typescript-eslint/no-explicit-any */

import type { SetupContext } from '@vue/composition-api';

export type Slots = SetupContext['slots'];

type CSSLength =
  | number
  | 'auto'
  | `${number}${'%' | 'px' | 'em' | 'rem' | 'vw' | 'vh'}`;

export enum SortDirection {
  /** 从小到大排序 */
  asc = 'ASC',
  /** 从大到小排序 */
  desc = 'DESC',
  /** 不参与排序 */
  none = 'NONE'
};

export type Column = {
  /** 显示的标题文本 */
  title: string;
  /** 映射到数据源内的 key */
  key: string;
  /** 内容插槽名，默认为 key 的值 */
  slot?: string;
  /** 标题插槽名，默认不使用插槽 */
  titleSlot?: string;
  /** 是否可排序 */
  sortable?: boolean;
  /** 当前的排序方向 */
  sortDirection?: SortDirection;
  /** 当前列宽 */
  width?: CSSLength;
};

export type TableDataType = Record<string, unknown>;

/** 提取部分 prop */
export function pick<TData extends TableDataType, TKey extends keyof TData>(
  data: TData,
  keys: TKey[]
): Pick<TData, TKey> {
    return ((Object.keys(data) as (keyof TData)[])
        .filter(key => (keys as (keyof TData)[]).includes(key)) as TKey[])
        .reduce((obj, key) => {
            obj[key] = data[key];
            return obj;
        }, {} as Pick<TData, TKey>);
};

/** 表格的 prop */
export const tableProps = {
  /** 显示表格边框 */
  border: {
    type: Boolean,
    default: true,
  },
  /** 数据源 */
  list: {
    type: Array as () => TableDataType[],
    default: () => [],
  },
  /** 列配置 */
  columns: {
    type: Array as () => Column[],
    required: true as const,
  },
  /** 分页配置 */
  pagerOptions: {
    type: Object as () => PagerPublicProps & { hidePager?: boolean },
  },
};

/** 翻页组件的 prop */
export const pagerProps = {
  /** 每页显示的数量 */
  limit: {
    type: Number,
    default: 20,
  },
  /** 总条数 */
  total: {
    type: Number,
    default: 0,
  },
  /** 当前页码 */
  page: {
    type: Number,
    default: 1,
  },
  /** 页码数量过多时，最多显示的页码数量 */
  maxShowPage: {
    type: Number,
    default: 7,
  },
};

/** 表格的 prop 类型 */
export type TablePublicProps = IxPublicPropTypes<typeof tableProps>;
/** 翻页组件的 prop 类型 */
export type PagerPublicProps = IxPublicPropTypes<typeof pagerProps>;
