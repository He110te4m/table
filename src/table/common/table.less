@borderWidth: 4px;
@borderTransparent: @borderWidth solid transparent;
@borderBlack: @borderWidth solid #000;

.table-sort-arrow {
  display: inline-block;
  vertical-align: middle;
  width: 0;
  height: 0;
  margin-left: 5px;
  opacity: 0.66;

  &.asc {
    border-left: @borderTransparent;
    border-right: @borderTransparent;
    border-bottom: @borderBlack;
  }

  &.desc {
    border-left: @borderTransparent;
    border-right: @borderTransparent;
    border-top: @borderBlack;
  }
}

.c-pager {
  display: flex;
  margin-top: 4px;

  &__item {
    padding: 4px 8px;
    border: 1px solid #e2e2e2;
    margin-right: 4px;
    border-radius: 2px;

    &:last-child {
        margin-right: 0;
    }

    &&--current {
        background-color: #333;
        color: #eee;
    }
  }
}
