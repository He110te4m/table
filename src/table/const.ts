export const TABLE_TOKEN = Symbol("CTable");
