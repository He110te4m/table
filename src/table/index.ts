/**
 * Created by uedc on 2021/10/11.
 */

import CTable from './Table'

export { CTable }
