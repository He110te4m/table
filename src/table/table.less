@borderWidth: 4px;
@borderTransparent: @borderWidth solid transparent;
@borderBlack: @borderWidth solid #000;

.table-sort-arrow {
  display: inline-block;
  vertical-align: middle;
  width: 0;
  height: 0;
  margin-left: 5px;
  opacity: 0.66;

  &.asc {
    border-left: @borderTransparent;
    border-right: @borderTransparent;
    border-bottom: @borderBlack;
  }

  &.desc {
    border-left: @borderTransparent;
    border-right: @borderTransparent;
    border-top: @borderBlack;
  }
}
