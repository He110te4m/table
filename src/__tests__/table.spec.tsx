import { mount, ThisTypedMountOptions } from '@vue/test-utils';
import { Column, SortDirection } from '../table/types';
import Vue from 'vue';
import { CTable } from '../table';

const columns: Column[] = [
  {
    title: 'name',
    key: 'name',
    slot: 'nameContent',
    titleSlot: 'nameSlot',
    sortable: true,
  },
  {
    title: 'age',
    key: 'age',
  },
];

describe('Table', () => {
  const TableMount = (options?: ThisTypedMountOptions<Vue>) =>
    mount(CTable, { ...options, propsData: { columns } });

  test('render', () => {
    const wrapper = TableMount();
    expect(wrapper.html()).toMatchSnapshot();
    expect(() => {
      wrapper.vm.$forceUpdate();
      wrapper.vm.$destroy();
    }).not.toThrow();
  });

  test('loadData', async () => {
    const wrapper = TableMount();
    await wrapper.setProps({
      list: [
        {
          name: '张三',
          age: 14,
        },
        {
          name: '李四',
          age: 16,
        },
      ],
    });
    expect(wrapper.html()).toMatchSnapshot();
    await wrapper.setProps({
      list: [
        {
          name: '张三',
          age: 18,
        },
        {
          name: '李四',
          age: 20,
        },
      ],
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  test('showPager', async () => {
    const wrapper = TableMount();
    await wrapper.setProps({
      list: [
        {
          name: '张三',
          age: 14,
        },
        {
          name: '李四',
          age: 16,
        },
      ],
      pagerOptions: {},
    });
    expect(wrapper.html()).toMatchSnapshot();

    await wrapper.setProps({
      pagerOptions: {
        total: 100,
        limit: 25,
        page: 1,
      },
    });
    expect(wrapper.html()).toMatchSnapshot();

    const allPageItem = wrapper.findAll('.c-pager__item__no');
    expect(allPageItem).toBeTruthy();

    const [, secondItem] = allPageItem.wrappers;
    expect(secondItem).toBeTruthy();

    await secondItem.trigger('click');
    expect(wrapper.html()).toMatchSnapshot();
  });

  test('sort', async () => {
    const wrapper = TableMount();
    await wrapper.setProps({
      list: [
        {
          name: '张三',
          age: 14,
        },
        {
          name: '李四',
          age: 16,
        },
      ],
    });
    expect(wrapper.html()).toMatchSnapshot();

    const headerItems = wrapper.findAll('.c-table-header__item');
    expect(headerItems).toBeTruthy();
    const [nameHeader] = headerItems.wrappers;
    expect(nameHeader).toBeTruthy();

    await nameHeader.trigger('click');
    expect(nameHeader.html()).toContain('table-sort-arrow');
    expect(nameHeader.html()).toMatchSnapshot();

    await nameHeader.trigger('click');
    expect(nameHeader.html()).toContain('table-sort-arrow');
    expect(nameHeader.html()).toMatchSnapshot();

    await nameHeader.trigger('click');
    expect(nameHeader.html()).not.toContain('table-sort-arrow');
    expect(nameHeader.html()).toMatchSnapshot();

    await nameHeader.trigger('click');
    expect(nameHeader.html()).toContain('table-sort-arrow');
    expect(nameHeader.html()).toMatchSnapshot();
  });

  test('default sort', async () => {
    const wrapper = TableMount();
    await wrapper.setProps({
      list: [
        {
          name: '张三',
          age: 14,
        },
        {
          name: '李四',
          age: 16,
        },
      ],
      columns: [
        {
          title: 'name',
          key: 'name',
          slot: 'nameContent',
          titleSlot: 'nameSlot',
          sortable: true,
          sortDirection: SortDirection.asc,
        },
        {
          title: 'age',
          key: 'age',
        },
      ],
    });
    expect(wrapper.html()).toMatchSnapshot();

    const headerItems = wrapper.findAll('.c-table-header__item');
    expect(headerItems).toBeTruthy();
    const [nameHeader] = headerItems.wrappers;
    expect(nameHeader).toBeTruthy();

    await nameHeader.trigger('click');
    expect(nameHeader.html()).toContain('table-sort-arrow');
    expect(nameHeader.html()).toMatchSnapshot();

    await nameHeader.trigger('click');
    expect(nameHeader.html()).not.toContain('table-sort-arrow');
    expect(nameHeader.html()).toMatchSnapshot();

    await nameHeader.trigger('click');
    expect(nameHeader.html()).toContain('table-sort-arrow');
    expect(nameHeader.html()).toMatchSnapshot();
  });

  test('sort not sortable column', async () => {
    const wrapper = TableMount();
    await wrapper.setProps({
      list: [
        {
          name: '张三',
          age: 14,
        },
        {
          name: '李四',
          age: 16,
        },
      ],
    });
    expect(wrapper.html()).toMatchSnapshot();

    const headerItems = wrapper.findAll('.c-table-header__item');
    expect(headerItems).toBeTruthy();
    const [, ageHeader] = headerItems.wrappers;
    expect(ageHeader).toBeTruthy();

    await ageHeader.trigger('click');
    expect(ageHeader.html()).not.toContain('table-sort-arrow');
    expect(ageHeader.html()).toMatchSnapshot();
  });

  test('title slot', async () => {
    const nameSlot = 'nameSlot';

    const wrapper = TableMount({
      slots: {
        nameSlot: nameSlot,
      },
    });
    await wrapper.setProps({
      list: [
        {
          name: '张三',
          age: 14,
        },
        {
          name: '李四',
          age: 16,
        },
      ],
    });
    expect(wrapper.html()).toMatchSnapshot();

    const headerItems = wrapper.findAll('.c-table-header__item');
    expect(headerItems).toBeTruthy();
    const [nameHeader] = headerItems.wrappers;
    expect(nameHeader).toBeTruthy();

    expect(nameHeader.text()).toEqual(nameSlot);
  });

  test('cell slot', async () => {
    const wrapper = TableMount({
      scopedSlots: {
        name: ({ value }: { value: string; }) => <span>Name is {value}</span>
      },
    });
    await wrapper.setProps({
      list: [
        {
          name: '张三',
          age: 14,
        },
        {
          name: '李四',
          age: 16,
        },
      ],
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  test('custom cell slot name', async () => {
    const wrapper = TableMount({
      scopedSlots: {
        nameContent: ({ value }: { value: string; }) => <span>Name is {value}</span>
      },
    });
    await wrapper.setProps({
      list: [
        {
          name: '张三',
          age: 14,
        },
        {
          name: '李四',
          age: 16,
        },
      ],
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  test('data missing', async () => {
    const wrapper = TableMount({
      scopedSlots: {
        nameContent: ({ value }: { value: string; }) => <span>Name is {value}</span>,
      },
    });
    await wrapper.setProps({
      list: [
        {
          age: 14,
        },
        {
          name: '李四',
        },
      ],
    });
    expect(wrapper.html()).toMatchSnapshot();
  });
  test('jump page method', () => {
    const wrapper = TableMount({
      propsData: [
        {
          name: '张三',
          age: 14,
        },
        {
          name: '李四',
          age: 16,
        },
      ]
    });
    expect(wrapper.html()).toMatchSnapshot();
    (wrapper.vm as Record<string, any>).jumpPage(2);
    expect(wrapper.html()).toMatchSnapshot();
  });
});
