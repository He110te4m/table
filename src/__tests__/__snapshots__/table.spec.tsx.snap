// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table cell slot 1`] = `
"<div class=\\"c-table-wrapper\\">
  <table border=\\"true\\" class=\\"c-table\\">
    <tr class=\\"c-table-header\\">
      <th class=\\"c-table-header__item\\"><span class=\\"c-table-header__item__text\\">name</span><span class=\\"table-sort-arrow asc c-table-header__item__icon\\"></span></th>
      <th class=\\"c-table-header__item\\"><span class=\\"c-table-header__item__text\\">age</span></th>
    </tr>
    <tbody class=\\"c-table-body\\">
      <tr class=\\"c-table-body__row\\">
        <td class=\\"c-table-body__row__cell\\"><span>Name is 李四</span></td>
        <td class=\\"c-table-body__row__cell\\">16</td>
      </tr>
      <tr class=\\"c-table-body__row\\">
        <td class=\\"c-table-body__row__cell\\"><span>Name is 张三</span></td>
        <td class=\\"c-table-body__row__cell\\">14</td>
      </tr>
    </tbody>
  </table>
  <dl class=\\"c-pager\\">
    <dd class=\\"c-pager__item c-pager__go-first\\"><a class=\\"c-pager__item__no\\">&lt;&lt;</a></dd>
    <dd class=\\"c-pager__item c-pager__item--current\\"><span class=\\"c-pager__item__no\\">1</span></dd>
    <dd class=\\"c-pager__item c-pager__go-end\\"><a class=\\"c-pager__item__no\\">&gt;&gt;</a></dd>
  </dl>
</div>"
`;

exports[`Table custom cell slot name 1`] = `
"<div class=\\"c-table-wrapper\\">
  <table border=\\"true\\" class=\\"c-table\\">
    <tr class=\\"c-table-header\\">
      <th class=\\"c-table-header__item\\"><span class=\\"c-table-header__item__text\\">name</span><span class=\\"table-sort-arrow asc c-table-header__item__icon\\"></span></th>
      <th class=\\"c-table-header__item\\"><span class=\\"c-table-header__item__text\\">age</span></th>
    </tr>
    <tbody class=\\"c-table-body\\">
      <tr class=\\"c-table-body__row\\">
        <td class=\\"c-table-body__row__cell\\"><span>Name is 李四</span></td>
        <td class=\\"c-table-body__row__cell\\">16</td>
      </tr>
      <tr class=\\"c-table-body__row\\">
        <td class=\\"c-table-body__row__cell\\"><span>Name is 张三</span></td>
        <td class=\\"c-table-body__row__cell\\">14</td>
      </tr>
    </tbody>
  </table>
  <dl class=\\"c-pager\\">
    <dd class=\\"c-pager__item c-pager__go-first\\"><a class=\\"c-pager__item__no\\">&lt;&lt;</a></dd>
    <dd class=\\"c-pager__item c-pager__item--current\\"><span class=\\"c-pager__item__no\\">1</span></dd>
    <dd class=\\"c-pager__item c-pager__go-end\\"><a class=\\"c-pager__item__no\\">&gt;&gt;</a></dd>
  </dl>
</div>"
`;

exports[`Table data missing 1`] = `
"<div class=\\"c-table-wrapper\\">
  <table border=\\"true\\" class=\\"c-table\\">
    <tr class=\\"c-table-header\\">
      <th class=\\"c-table-header__item\\"><span class=\\"c-table-header__item__text\\">name</span><span class=\\"table-sort-arrow asc c-table-header__item__icon\\"></span></th>
      <th class=\\"c-table-header__item\\"><span class=\\"c-table-header__item__text\\">age</span></th>
    </tr>
    <tbody class=\\"c-table-body\\">
      <tr class=\\"c-table-body__row\\">
        <td class=\\"c-table-body__row__cell\\"><span>Name is </span></td>
        <td class=\\"c-table-body__row__cell\\">14</td>
      </tr>
      <tr class=\\"c-table-body__row\\">
        <td class=\\"c-table-body__row__cell\\"><span>Name is 李四</span></td>
        <td class=\\"c-table-body__row__cell\\"></td>
      </tr>
    </tbody>
  </table>
  <dl class=\\"c-pager\\">
    <dd class=\\"c-pager__item c-pager__go-first\\"><a class=\\"c-pager__item__no\\">&lt;&lt;</a></dd>
    <dd class=\\"c-pager__item c-pager__item--current\\"><span class=\\"c-pager__item__no\\">1</span></dd>
    <dd class=\\"c-pager__item c-pager__go-end\\"><a class=\\"c-pager__item__no\\">&gt;&gt;</a></dd>
  </dl>
</div>"
`;

exports[`Table default sort 1`] = `
"<div class=\\"c-table-wrapper\\">
  <table border=\\"true\\" class=\\"c-table\\">
    <tr class=\\"c-table-header\\">
      <th class=\\"c-table-header__item\\"><span class=\\"c-table-header__item__text\\">name</span><span class=\\"table-sort-arrow asc c-table-header__item__icon\\"></span></th>
      <th class=\\"c-table-header__item\\"><span class=\\"c-table-header__item__text\\">age</span></th>
    </tr>
    <tbody class=\\"c-table-body\\">
      <tr class=\\"c-table-body__row\\">
        <td class=\\"c-table-body__row__cell\\">李四</td>
        <td class=\\"c-table-body__row__cell\\">16</td>
      </tr>
      <tr class=\\"c-table-body__row\\">
        <td class=\\"c-table-body__row__cell\\">张三</td>
        <td class=\\"c-table-body__row__cell\\">14</td>
      </tr>
    </tbody>
  </table>
  <dl class=\\"c-pager\\">
    <dd class=\\"c-pager__item c-pager__go-first\\"><a class=\\"c-pager__item__no\\">&lt;&lt;</a></dd>
    <dd class=\\"c-pager__item c-pager__item--current\\"><span class=\\"c-pager__item__no\\">1</span></dd>
    <dd class=\\"c-pager__item c-pager__go-end\\"><a class=\\"c-pager__item__no\\">&gt;&gt;</a></dd>
  </dl>
</div>"
`;

exports[`Table default sort 2`] = `"<th class=\\"c-table-header__item\\"><span class=\\"c-table-header__item__text\\">name</span><span class=\\"table-sort-arrow desc c-table-header__item__icon\\"></span></th>"`;

exports[`Table default sort 3`] = `"<th class=\\"c-table-header__item\\"><span class=\\"c-table-header__item__text\\">name</span></th>"`;

exports[`Table default sort 4`] = `"<th class=\\"c-table-header__item\\"><span class=\\"c-table-header__item__text\\">name</span><span class=\\"table-sort-arrow asc c-table-header__item__icon\\"></span></th>"`;

exports[`Table loadData 1`] = `
"<div class=\\"c-table-wrapper\\">
  <table border=\\"true\\" class=\\"c-table\\">
    <tr class=\\"c-table-header\\">
      <th class=\\"c-table-header__item\\"><span class=\\"c-table-header__item__text\\">name</span></th>
      <th class=\\"c-table-header__item\\"><span class=\\"c-table-header__item__text\\">age</span></th>
    </tr>
    <tbody class=\\"c-table-body\\">
      <tr class=\\"c-table-body__row\\">
        <td class=\\"c-table-body__row__cell\\">张三</td>
        <td class=\\"c-table-body__row__cell\\">14</td>
      </tr>
      <tr class=\\"c-table-body__row\\">
        <td class=\\"c-table-body__row__cell\\">李四</td>
        <td class=\\"c-table-body__row__cell\\">16</td>
      </tr>
    </tbody>
  </table>
  <dl class=\\"c-pager\\">
    <dd class=\\"c-pager__item c-pager__go-first\\"><a class=\\"c-pager__item__no\\">&lt;&lt;</a></dd>
    <dd class=\\"c-pager__item c-pager__item--current\\"><span class=\\"c-pager__item__no\\">1</span></dd>
    <dd class=\\"c-pager__item c-pager__go-end\\"><a class=\\"c-pager__item__no\\">&gt;&gt;</a></dd>
  </dl>
</div>"
`;

exports[`Table loadData 2`] = `
"<div class=\\"c-table-wrapper\\">
  <table border=\\"true\\" class=\\"c-table\\">
    <tr class=\\"c-table-header\\">
      <th class=\\"c-table-header__item\\"><span class=\\"c-table-header__item__text\\">name</span></th>
      <th class=\\"c-table-header__item\\"><span class=\\"c-table-header__item__text\\">age</span></th>
    </tr>
    <tbody class=\\"c-table-body\\">
      <tr class=\\"c-table-body__row\\">
        <td class=\\"c-table-body__row__cell\\">张三</td>
        <td class=\\"c-table-body__row__cell\\">18</td>
      </tr>
      <tr class=\\"c-table-body__row\\">
        <td class=\\"c-table-body__row__cell\\">李四</td>
        <td class=\\"c-table-body__row__cell\\">20</td>
      </tr>
    </tbody>
  </table>
  <dl class=\\"c-pager\\">
    <dd class=\\"c-pager__item c-pager__go-first\\"><a class=\\"c-pager__item__no\\">&lt;&lt;</a></dd>
    <dd class=\\"c-pager__item c-pager__item--current\\"><span class=\\"c-pager__item__no\\">1</span></dd>
    <dd class=\\"c-pager__item c-pager__go-end\\"><a class=\\"c-pager__item__no\\">&gt;&gt;</a></dd>
  </dl>
</div>"
`;

exports[`Table render 1`] = `
"<div class=\\"c-table-wrapper\\">
  <table border=\\"true\\" class=\\"c-table\\">
    <tr class=\\"c-table-header\\">
      <th class=\\"c-table-header__item\\"><span class=\\"c-table-header__item__text\\">name</span></th>
      <th class=\\"c-table-header__item\\"><span class=\\"c-table-header__item__text\\">age</span></th>
    </tr>
    <tbody class=\\"c-table-body\\"></tbody>
  </table>
  <dl class=\\"c-pager\\">
    <dd class=\\"c-pager__item c-pager__go-first\\"><a class=\\"c-pager__item__no\\">&lt;&lt;</a></dd>
    <dd class=\\"c-pager__item c-pager__go-end\\"><a class=\\"c-pager__item__no\\">&gt;&gt;</a></dd>
  </dl>
</div>"
`;

exports[`Table showPager 1`] = `
"<div class=\\"c-table-wrapper\\">
  <table border=\\"true\\" class=\\"c-table\\">
    <tr class=\\"c-table-header\\">
      <th class=\\"c-table-header__item\\"><span class=\\"c-table-header__item__text\\">name</span></th>
      <th class=\\"c-table-header__item\\"><span class=\\"c-table-header__item__text\\">age</span></th>
    </tr>
    <tbody class=\\"c-table-body\\">
      <tr class=\\"c-table-body__row\\">
        <td class=\\"c-table-body__row__cell\\">张三</td>
        <td class=\\"c-table-body__row__cell\\">14</td>
      </tr>
      <tr class=\\"c-table-body__row\\">
        <td class=\\"c-table-body__row__cell\\">李四</td>
        <td class=\\"c-table-body__row__cell\\">16</td>
      </tr>
    </tbody>
  </table>
  <dl class=\\"c-pager\\">
    <dd class=\\"c-pager__item c-pager__go-first\\"><a class=\\"c-pager__item__no\\">&lt;&lt;</a></dd>
    <dd class=\\"c-pager__item c-pager__item--current\\"><span class=\\"c-pager__item__no\\">1</span></dd>
    <dd class=\\"c-pager__item c-pager__go-end\\"><a class=\\"c-pager__item__no\\">&gt;&gt;</a></dd>
  </dl>
</div>"
`;

exports[`Table showPager 2`] = `
"<div class=\\"c-table-wrapper\\">
  <table border=\\"true\\" class=\\"c-table\\">
    <tr class=\\"c-table-header\\">
      <th class=\\"c-table-header__item\\"><span class=\\"c-table-header__item__text\\">name</span></th>
      <th class=\\"c-table-header__item\\"><span class=\\"c-table-header__item__text\\">age</span></th>
    </tr>
    <tbody class=\\"c-table-body\\">
      <tr class=\\"c-table-body__row\\">
        <td class=\\"c-table-body__row__cell\\">张三</td>
        <td class=\\"c-table-body__row__cell\\">14</td>
      </tr>
      <tr class=\\"c-table-body__row\\">
        <td class=\\"c-table-body__row__cell\\">李四</td>
        <td class=\\"c-table-body__row__cell\\">16</td>
      </tr>
    </tbody>
  </table>
  <dl class=\\"c-pager\\">
    <dd class=\\"c-pager__item c-pager__go-first\\"><a class=\\"c-pager__item__no\\">&lt;&lt;</a></dd>
    <dd class=\\"c-pager__item c-pager__item--current\\"><span class=\\"c-pager__item__no\\">1</span></dd>
    <dd class=\\"c-pager__item \\"><a class=\\"c-pager__item__no\\">2</a></dd>
    <dd class=\\"c-pager__item \\"><a class=\\"c-pager__item__no\\">3</a></dd>
    <dd class=\\"c-pager__item \\"><a class=\\"c-pager__item__no\\">4</a></dd>
    <dd class=\\"c-pager__item c-pager__go-end\\"><a class=\\"c-pager__item__no\\">&gt;&gt;</a></dd>
  </dl>
</div>"
`;

exports[`Table showPager 3`] = `
"<div class=\\"c-table-wrapper\\">
  <table border=\\"true\\" class=\\"c-table\\">
    <tr class=\\"c-table-header\\">
      <th class=\\"c-table-header__item\\"><span class=\\"c-table-header__item__text\\">name</span></th>
      <th class=\\"c-table-header__item\\"><span class=\\"c-table-header__item__text\\">age</span></th>
    </tr>
    <tbody class=\\"c-table-body\\">
      <tr class=\\"c-table-body__row\\">
        <td class=\\"c-table-body__row__cell\\">张三</td>
        <td class=\\"c-table-body__row__cell\\">14</td>
      </tr>
      <tr class=\\"c-table-body__row\\">
        <td class=\\"c-table-body__row__cell\\">李四</td>
        <td class=\\"c-table-body__row__cell\\">16</td>
      </tr>
    </tbody>
  </table>
  <dl class=\\"c-pager\\">
    <dd class=\\"c-pager__item c-pager__go-first\\"><a class=\\"c-pager__item__no\\">&lt;&lt;</a></dd>
    <dd class=\\"c-pager__item c-pager__item--current\\"><span class=\\"c-pager__item__no\\">1</span></dd>
    <dd class=\\"c-pager__item \\"><a class=\\"c-pager__item__no\\">2</a></dd>
    <dd class=\\"c-pager__item \\"><a class=\\"c-pager__item__no\\">3</a></dd>
    <dd class=\\"c-pager__item \\"><a class=\\"c-pager__item__no\\">4</a></dd>
    <dd class=\\"c-pager__item c-pager__go-end\\"><a class=\\"c-pager__item__no\\">&gt;&gt;</a></dd>
  </dl>
</div>"
`;

exports[`Table sort 1`] = `
"<div class=\\"c-table-wrapper\\">
  <table border=\\"true\\" class=\\"c-table\\">
    <tr class=\\"c-table-header\\">
      <th class=\\"c-table-header__item\\"><span class=\\"c-table-header__item__text\\">name</span></th>
      <th class=\\"c-table-header__item\\"><span class=\\"c-table-header__item__text\\">age</span></th>
    </tr>
    <tbody class=\\"c-table-body\\">
      <tr class=\\"c-table-body__row\\">
        <td class=\\"c-table-body__row__cell\\">张三</td>
        <td class=\\"c-table-body__row__cell\\">14</td>
      </tr>
      <tr class=\\"c-table-body__row\\">
        <td class=\\"c-table-body__row__cell\\">李四</td>
        <td class=\\"c-table-body__row__cell\\">16</td>
      </tr>
    </tbody>
  </table>
  <dl class=\\"c-pager\\">
    <dd class=\\"c-pager__item c-pager__go-first\\"><a class=\\"c-pager__item__no\\">&lt;&lt;</a></dd>
    <dd class=\\"c-pager__item c-pager__item--current\\"><span class=\\"c-pager__item__no\\">1</span></dd>
    <dd class=\\"c-pager__item c-pager__go-end\\"><a class=\\"c-pager__item__no\\">&gt;&gt;</a></dd>
  </dl>
</div>"
`;

exports[`Table sort 2`] = `"<th class=\\"c-table-header__item\\"><span class=\\"c-table-header__item__text\\">name</span><span class=\\"table-sort-arrow asc c-table-header__item__icon\\"></span></th>"`;

exports[`Table sort 3`] = `"<th class=\\"c-table-header__item\\"><span class=\\"c-table-header__item__text\\">name</span><span class=\\"table-sort-arrow desc c-table-header__item__icon\\"></span></th>"`;

exports[`Table sort 4`] = `"<th class=\\"c-table-header__item\\"><span class=\\"c-table-header__item__text\\">name</span></th>"`;

exports[`Table sort 5`] = `"<th class=\\"c-table-header__item\\"><span class=\\"c-table-header__item__text\\">name</span><span class=\\"table-sort-arrow asc c-table-header__item__icon\\"></span></th>"`;

exports[`Table sort not sortable column 1`] = `
"<div class=\\"c-table-wrapper\\">
  <table border=\\"true\\" class=\\"c-table\\">
    <tr class=\\"c-table-header\\">
      <th class=\\"c-table-header__item\\"><span class=\\"c-table-header__item__text\\">name</span><span class=\\"table-sort-arrow asc c-table-header__item__icon\\"></span></th>
      <th class=\\"c-table-header__item\\"><span class=\\"c-table-header__item__text\\">age</span></th>
    </tr>
    <tbody class=\\"c-table-body\\">
      <tr class=\\"c-table-body__row\\">
        <td class=\\"c-table-body__row__cell\\">李四</td>
        <td class=\\"c-table-body__row__cell\\">16</td>
      </tr>
      <tr class=\\"c-table-body__row\\">
        <td class=\\"c-table-body__row__cell\\">张三</td>
        <td class=\\"c-table-body__row__cell\\">14</td>
      </tr>
    </tbody>
  </table>
  <dl class=\\"c-pager\\">
    <dd class=\\"c-pager__item c-pager__go-first\\"><a class=\\"c-pager__item__no\\">&lt;&lt;</a></dd>
    <dd class=\\"c-pager__item c-pager__item--current\\"><span class=\\"c-pager__item__no\\">1</span></dd>
    <dd class=\\"c-pager__item c-pager__go-end\\"><a class=\\"c-pager__item__no\\">&gt;&gt;</a></dd>
  </dl>
</div>"
`;

exports[`Table sort not sortable column 2`] = `"<th class=\\"c-table-header__item\\"><span class=\\"c-table-header__item__text\\">age</span></th>"`;

exports[`Table sortMethod 1`] = `
"<div class=\\"c-table-wrapper\\">
  <table border=\\"true\\" class=\\"c-table\\">
    <tr class=\\"c-table-header\\">
      <th class=\\"c-table-header__item\\"><span class=\\"c-table-header__item__text\\">name</span><span class=\\"table-sort-arrow asc c-table-header__item__icon\\"></span></th>
      <th class=\\"c-table-header__item\\"><span class=\\"c-table-header__item__text\\">age</span></th>
    </tr>
    <tbody class=\\"c-table-body\\"></tbody>
  </table>
  <dl class=\\"c-pager\\">
    <dd class=\\"c-pager__item c-pager__go-first\\"><a class=\\"c-pager__item__no\\">&lt;&lt;</a></dd>
    <dd class=\\"c-pager__item c-pager__go-end\\"><a class=\\"c-pager__item__no\\">&gt;&gt;</a></dd>
  </dl>
</div>"
`;

exports[`Table sortMethod 2`] = `
"<div class=\\"c-table-wrapper\\">
  <table border=\\"true\\" class=\\"c-table\\">
    <tr class=\\"c-table-header\\">
      <th class=\\"c-table-header__item\\"><span class=\\"c-table-header__item__text\\">name</span><span class=\\"table-sort-arrow asc c-table-header__item__icon\\"></span></th>
      <th class=\\"c-table-header__item\\"><span class=\\"c-table-header__item__text\\">age</span></th>
    </tr>
    <tbody class=\\"c-table-body\\"></tbody>
  </table>
  <dl class=\\"c-pager\\">
    <dd class=\\"c-pager__item c-pager__go-first\\"><a class=\\"c-pager__item__no\\">&lt;&lt;</a></dd>
    <dd class=\\"c-pager__item c-pager__go-end\\"><a class=\\"c-pager__item__no\\">&gt;&gt;</a></dd>
  </dl>
</div>"
`;

exports[`Table title slot 1`] = `
"<div class=\\"c-table-wrapper\\">
  <table border=\\"true\\" class=\\"c-table\\">
    <tr class=\\"c-table-header\\">
      <th class=\\"c-table-header__item\\"><span class=\\"c-table-header__item__text\\">nameSlot</span><span class=\\"table-sort-arrow asc c-table-header__item__icon\\"></span></th>
      <th class=\\"c-table-header__item\\"><span class=\\"c-table-header__item__text\\">age</span></th>
    </tr>
    <tbody class=\\"c-table-body\\">
      <tr class=\\"c-table-body__row\\">
        <td class=\\"c-table-body__row__cell\\">李四</td>
        <td class=\\"c-table-body__row__cell\\">16</td>
      </tr>
      <tr class=\\"c-table-body__row\\">
        <td class=\\"c-table-body__row__cell\\">张三</td>
        <td class=\\"c-table-body__row__cell\\">14</td>
      </tr>
    </tbody>
  </table>
  <dl class=\\"c-pager\\">
    <dd class=\\"c-pager__item c-pager__go-first\\"><a class=\\"c-pager__item__no\\">&lt;&lt;</a></dd>
    <dd class=\\"c-pager__item c-pager__item--current\\"><span class=\\"c-pager__item__no\\">1</span></dd>
    <dd class=\\"c-pager__item c-pager__go-end\\"><a class=\\"c-pager__item__no\\">&gt;&gt;</a></dd>
  </dl>
</div>"
`;
